### DISCUSSION TITLE
Using child_process during Docusaurus build

### DISCUSSION THREAD

**Discussion by vvolhejn** [facebook/docusaurus#8497] [Q&A] [answered]
Hi! I'm trying to add a Python API reference to our Docusaurus docs page. I have a Python script (based on [MkDocstrings](https://mkdocstrings.github.io/)) that I can give the name of a Python module/function and it will render its documentation in HTML. I'm trying to add a MDX tag that I could use like `<PythonDoc name="collections.OrderedDict" />` to insert the Python documentation. I tried running this script from 
```js
import {spawnSync} from "child_process";

...

export default function PythonDoc({name}: { name: string }): Node {
  const result = spawnSync(
    ".../render_symbol.py",
    ["collections.OrderedDict"], // args
  )
  // ... do something
}
```
but when I run `yarn run build`, I get
```
Module not found: Error: Can't resolve 'child_process' in '/Users/vaclav/prog/lakera/documentation/src'
[ERROR] Client bundle compiled with errors therefore further build is impossible.
```
My understanding is that this is because the code is "meant" to be client-side so running subprocesses is not an option. However, I am actually only running server-side to get static HTML, which I then serve to the client – so using subprocesses "should" be ok. Is there a way to get around this limitation? Or would a different solution altogether be better?

I also looked at [creating a Remark plugin](https://docusaurus.io/docs/markdown-features/plugins#creating-new-rehyperemark-plugins) which might do the job, but working directly with the Markdown/HTML AST rather than MDX is more difficult.

Edit: A hacky solution is to make a Python script that will preprocess the Markdown files, look for lines matching a defined pattern like `:::python collections.OrderedDict`, and replace them with the rendered content, even before any Docusaurus processing is applied. But this requires creating separate "template" markdown files. I still wonder whether there is a more robust way to do this.

**Comment by slorber** [comment, marked as answer]
It is really hard for me to understand what you are trying to do because you don't really provide me all the info.

What I can tell is that indeed it's not ok to use Node.js code in the browser (ie in a React theme component), and there's no way to work around that.

That's not really optimized, but you could create a plugin that uses the `setGlobalData` API and create all the data required to make your client-side API work (I have no idea what this data looks like, JSON, HTML?). Then in the component you can do `usePluginData()["collections.OrderedDict"]` to retrieve it. It is not optimized because this data will be loaded on all your pages of your sites, so if it's 10mb that's a lot of data to load for all pages (some of them do not even display that data).

Building a Remark plugin looks like a good solution indeed. With MDX 2 (upgrade in progress) you could use the remark-directives to invent easily your own syntax: https://github.com/remarkjs/remark-directive (but you still need to tell remark what to do with that syntax, there are some good examples online and it's quite similar to what we do with admonitions)

**Reply by vvolhejn** [reply]
Thank you for the answer! Sorry, I realize after re-reading that I didn't explain myself very well.

Ultimately, I want the to enable the doc writers to insert auto-generated Python API reference from Markdown. This should take the form of a directive like `:::python collections.OrderedDict` (or whatever different syntax) that people can put into their Markdown to get the reference for that symbol.

To do this, I have a script `render_symbol.py` that I can run like
```sh
python render_symbol.py collections.OrderedDict
```
and it will output the symbol's documentation as HTML, e.g.
```html
<p>
Return an instance of a <pre>dict</pre> subclass that has methods specialized for rearranging dictionary order.
</p>
(...)
```

Now the question is how to turn that script into a Markdown directive. This is the problem I was trying to solve. The candidate solutions were
a) a React `<PythonDoc/>` component - you ruled this impossible
b) a Remark plugin
c) hacky solution where I just preprocess the files and insert the API reference before Docusaurus is even run

The documentation deployed by serving the static files produced by `docusaurus build`.

In the meantime I went with solution c), but it has some drawbacks, so in the future I'll try to switch to b). I wasn't aware of `setGlobalData`, thank you! I think the data might be around 100kB now, but I'm expecting there to be maybe 1MB or so in the future – that's already fairly for something that's not always used.

**Reply by slorber** [reply]
100kb seems acceptable for global data will still slow down your site (ie your homepage would need to lead that data too).

Using a remark plugin and replacing things directly in the markdown document is likely to be a more efficient solution.

Until mdx2 + directives, you can try to use a markdown variable substitution plugin, like discussed here: https://github.com/facebook/docusaurus/issues/395
